### DISCUSSION TITLE
Pattern for inserts into tables that have many-to-one relation

### DISCUSSION THREAD

**Discussion by lovetox** [sqlalchemy/sqlalchemy#11042] [Usage Questions] [answered]
Imagine a mail program which can store thousands of messages that are all from the same mail address

My idea was, less storage is used if i save the mail address only once, and reference it in the message with a foreign key

something like this

```python
class Address(Base):
    __tablename__ = "address"

    entitykey: Mapped[int] = mapped_column(primary_key=True)
    address: Mapped[str] = mapped_column(unique=True)


class Message(Base):
    __tablename__ = "message"

    entitykey: Mapped[int] = mapped_column(primary_key=True)
    fk_address_entitykey: Mapped[int] = mapped_column(ForeignKey("address.entitykey"))
    text: Mapped[str]

    address: Mapped[Address] = relationship()
```

if i receive a mail message, and i create a Message Object which i want to insert 

```python
session.add(
    Message(text="hi",  address=Address(address="mymail@mail.com"))
)
```

this will work the first time, but on the second message it will fail because of the unique constraint.
But i feel this must be very common case in the relational database world.

What is a good approach here? Do i really always have to query first if the object exists? Can SQLAlchemy make this easier somewhow?

Thanks

**Comment by gordthompson** [comment, marked as answer]
related: https://github.com/sqlalchemy/sqlalchemy/discussions/11028
